### PR TITLE
Enable Sglang diffusion on Intel XPU

### PR DESCRIPTION
**Pull request by sushildubey171**
## Motivation

To add support for running diffusion models on Intel XPU devices

## Modifications

- **Introduce Intel XPU platform + registration**
  - Implemented `XpuPlatform` (device info, memory queries, attention backend selection, communicator stub)
  - Added XPU detection plugin and platform resolution logic
  - Extended platform interface to include `XPU`, device selection, distributed backend string, and device-module seeding

- **Distributed + device selection updates for XPU**
  - Defaulted distributed backend to platform-selected backend (instead of hardcoding NCCL)
  - Avoided passing `device_id` to `init_process_group` for XPU (similar to MPS/MUSA handling)

- **Make GPU worker device-agnostic (CUDA/XPU)**

- **Add XPU fallbacks for custom ops / kernels**
  - Added `forward_xpu()` paths that fall back to native PyTorch implementations
  - Updated Triton ops to accept CUDA **or** XPU tensors; made warp-size detection safe when CUDA is unavailable

- **Model-level adjustments for XPU**
  - Used a non-flashinfer rotary embedding path on XPU
  - Avoided `is_causal=True` + `attn_mask` combination on XPU (same handling as ROCm)

## Accuracy Tests

We tested flux.1-dev model for image generation, visually it looks good. Also verified FID score on subset of coco dataset (2k samples) and it is comparable with cuda. FID score 50.8 vs 51.1.  

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/pyproject_xpu.toml [hunk 1]
```diff
@@ -73,6 +73,18 @@ dependencies = [
...
+diffusion = [
+  "PyYAML==6.0.1",
+  "cloudpickle",
+  "diffusers==0.35.2",
+  "imageio==2.36.0",
+  "imageio-ffmpeg==0.5.1",
+  "moviepy>=2.0.0",
+  "opencv-python==4.10.0.84",
+  "remote-pdb",
+  "st_attn ==0.0.7",
```
Comment (msinnha1) [inline python/pyproject_xpu.toml]: typo error, please remove extra space in "st_attn ==0.0.7"

File: python/sglang/multimodal_gen/runtime/utils/profiler.py [hunk 2]
```diff
@@ -43,6 +43,9 @@ def __init__(
         if torch.cuda.is_available():
             activities.append(torch.profiler.ProfilerActivity.CUDA)
 
+        if torch.xpu.is_available():
```
Comment (msinnha1) [inline python/sglang/multimodal_gen/runtime/utils/profiler.py]: Minor comment: this should be:

if hasattr(torch, "xpu") and torch.xpu.is_available():

**Review by msinnha1 — COMMENTED** [review]
majority of things looks good to me!
But please add either the test ran results or some basic benchmarking numbers with the PR, even if it is a for a reduced configuration.

File: python/sglang/multimodal_gen/runtime/platforms/xpu.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    _enum = PlatformEnum.XPU
+    device_name: str = "xpu"
+    device_type: str = "xpu"
+    dispatch_key: str = "XPU"
+    device_control_env_var: str = "ZE_AFFINITY_MASK"
+
+    @classmethod
+    def get_device_capability(cls, device_id: int = 0) -> DeviceCapability | None:
+        """Intel XPU doesn't have CUDA-style compute capability."""
+        return None
```
Comment (mingfeima) [inline python/sglang/multimodal_gen/runtime/platforms/xpu.py]: @airMeng how we are aligning this with pytorch side?

File: python/sglang/multimodal_gen/runtime/platforms/xpu.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        cls,
+        selected_backend: AttentionBackendEnum | None,
+        head_size: int,
+        dtype: torch.dtype,
+    ) -> str:
+        """Get the attention backend class string for Intel XPU.
+
+        Intel XPU supports PyTorch's native SDPA which is the most portable option.
+        """
+        if (
```
Comment (mingfeima) [inline python/sglang/multimodal_gen/runtime/platforms/xpu.py]: using `TORCH.SPDA` is not good choice here, it is basically a reference design.

suggest that we use XPU attn backend directly, the interface is samiliar to FA backend.

SDPA backend will be troublesome when you really have `varlen`.

additionally, we are very likely to use different optimizaton scheme than torch.SDPA, can't rely perf on it, the release cycle for torch is too long.

**Review by mingfeima — CHANGES_REQUESTED** [review]

**Comment by mingfeima** [comment]
On top of everything, this one is not in our KPI list yet, anyway it is still nice to have as long as our prioritized tasks are on track :)

This one mostly looks fine, the biggest issue is attention backend. Aside from this, we need more detailed performance benchmark results and profiling of hotspots.

File: python/sglang/multimodal_gen/runtime/platforms/xpu.py (hunk 3 above)
Comment (airMeng) [reply python/sglang/multimodal_gen/runtime/platforms/xpu.py]: please use functions added in https://github.com/sgl-project/sgl-kernel-xpu/pull/101

File: python/sglang/multimodal_gen/runtime/layers/attention/backends/xpu_backend.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+        # TODO: put empty values here to be set at first-run, since the q_len calculation can be complicated
+        return FlashAttentionMetadata(max_seqlen_q=None, max_seqlen_k=None)
+
+
+class XPUAttentionBackend(AttentionBackend):
+    accept_output_buffer: bool = True
+
+    @staticmethod
+    def get_supported_head_sizes() -> list[int]:
+        return [32, 64, 96, 128, 160, 192, 224, 256]
```
Comment (polisettyvarma) [inline python/sglang/multimodal_gen/runtime/layers/attention/backends/xpu_backend.py]: remove 32, 160 and 224

File: python/sglang/multimodal_gen/runtime/layers/attention/backends/xpu_backend.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+    def get_enum() -> AttentionBackendEnum:
+        return AttentionBackendEnum.XPU_FA
+
+    @staticmethod
+    def get_impl_cls() -> type["XPUAttentionImpl"]:
+        return XPUAttentionImpl
+
+    @staticmethod
+    def get_metadata_cls() -> type["AttentionMetadata"]:
+        raise NotImplementedError
```
Comment (polisettyvarma) [inline python/sglang/multimodal_gen/runtime/layers/attention/backends/xpu_backend.py]: why can't we return our class ?
some backends do and don't

File: python/sglang/multimodal_gen/runtime/layers/layernorm.py [hunk 7]
```diff
@@ -10,6 +10,17 @@
...
+from sglang.multimodal_gen.runtime.platforms import current_platform
+
+_is_cuda = current_platform.is_cuda()
+_is_npu = current_platform.is_npu()
+_is_xpu = current_platform.is_xpu()
+if _is_cuda or _is_xpu:
+    from sgl_kernel import fused_add_rmsnorm, rmsnorm
+
+if _is_npu:
+    import torch_npu
```
Comment (polisettyvarma) [inline python/sglang/multimodal_gen/runtime/layers/layernorm.py]: intentional ?

**Review by polisettyvarma — CHANGES_REQUESTED** [review]

File: python/sglang/multimodal_gen/runtime/layers/layernorm.py (hunk 7 above)
Comment (sushildubey171) [reply python/sglang/multimodal_gen/runtime/layers/layernorm.py]: this is typo, it should be for xpu

File: python/sglang/multimodal_gen/runtime/layers/attention/backends/xpu_backend.py (hunk 6 above)
Comment (sushildubey171) [reply python/sglang/multimodal_gen/runtime/layers/attention/backends/xpu_backend.py]: I have removed it, not required for us

File: python/sglang/multimodal_gen/runtime/layers/attention/backends/xpu_backend.py (hunk 5 above)
Comment (sushildubey171) [reply python/sglang/multimodal_gen/runtime/layers/attention/backends/xpu_backend.py]: ok

File: python/sglang/multimodal_gen/runtime/platforms/xpu.py (hunk 4 above)
Comment (sushildubey171) [reply python/sglang/multimodal_gen/runtime/platforms/xpu.py]: integrated the attention backend

File: python/sglang/multimodal_gen/configs/models/dits/base.py [hunk 8]
```diff
@@ -25,6 +25,7 @@ class DiTArchConfig(ArchConfig):
     reverse_param_names_mapping: dict = field(default_factory=dict)
     _supported_attention_backends: set[AttentionBackendEnum] = field(
         default_factory=lambda: {
+            AttentionBackendEnum.XPU_FA,
```
Comment (mickqian) [inline python/sglang/multimodal_gen/configs/models/dits/base.py]: would it be possible we still use `FA`, but handles them differently when resolving on different platforms?

Comment (sushildubey171) [reply python/sglang/multimodal_gen/configs/models/dits/base.py]: sure, thanks for your review @mickqian, I have fixed it.

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py [hunk 9]
```diff
@@ -1498,6 +1498,9 @@ def _build_attn_metadata(
                 }
             )
         elif self.attn_backend.get_enum() == AttentionBackendEnum.FA:
+            if current_platform.is_xpu():
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py]: could you avoid `current_platform.is_xpu` call from common files, please consider platform design

Comment (sushildubey171) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py]: ok, done

File: python/sglang/multimodal_gen/runtime/utils/profiler.py (hunk 2 above)
Comment (sushildubey171) [reply python/sglang/multimodal_gen/runtime/utils/profiler.py]: done

**Review by mickqian — APPROVED** [review]

**Review by polisettyvarma — APPROVED** [review]

**Review by mingfeima — APPROVED** [review]
